extensions/wrapper/wrapper/src/main/java/de/sovity/edc/ext/wrapper/api/ui/pages/transferhistory/TransferHistoryPageDataFetcher.java

```diff
@@ -0,0 +1,136 @@
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       sovity GmbH - initial API and implementation
+ *
+ */
+
+package de.sovity.edc.ext.wrapper.api.ui.pages.transferhistory;
+
+import de.sovity.edc.ext.wrapper.api.ui.model.ContractAgreementDirection;
+import de.sovity.edc.ext.wrapper.api.ui.model.TransferHistoryEntry;
+import de.sovity.edc.ext.wrapper.api.ui.pages.contracts.services.TransferProcessStateService;
+import lombok.RequiredArgsConstructor;
+import org.eclipse.edc.connector.contract.spi.negotiation.store.ContractNegotiationStore;
+import org.eclipse.edc.connector.contract.spi.types.agreement.ContractAgreement;
+import org.eclipse.edc.connector.contract.spi.types.negotiation.ContractNegotiation;
+import org.eclipse.edc.connector.spi.asset.AssetService;
+import org.eclipse.edc.connector.spi.contractagreement.ContractAgreementService;
+import org.eclipse.edc.connector.spi.transferprocess.TransferProcessService;
+import org.eclipse.edc.connector.transfer.spi.types.TransferProcess;
+import org.eclipse.edc.spi.entity.Entity;
+import org.eclipse.edc.spi.query.QuerySpec;
+import org.eclipse.edc.spi.types.domain.asset.Asset;
+import org.jetbrains.annotations.NotNull;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import static de.sovity.edc.ext.wrapper.utils.EdcDateUtils.utcMillisToOffsetDateTime;
+import static java.util.stream.Collectors.*;
+
+@RequiredArgsConstructor
+public class TransferHistoryPageDataFetcher {
+    private final AssetService assetService;
+    private final ContractAgreementService contractAgreementService;
+    private final ContractNegotiationStore contractNegotiationStore;
+    private final TransferProcessService transferProcessService;
+
+    private final TransferProcessStateService transferProcessStateService;
+
+    /**
+     * Fetches all contract agreements as {@link TransferHistoryEntry}s.
+     *
+     * @return {@link TransferHistoryEntry}s
+     */
+    @NotNull
+    public List<TransferHistoryEntry> getTransferHistoryEntries() {
+
+
+        var negotiationsByID = getAllContractNegotiations().stream()
+                .filter(it -> it.getContractAgreement() != null)
+                .collect(groupingBy(
+                        it -> it.getContractAgreement().getId(),
+                        collectingAndThen(maxBy(Comparator.comparing(Entity::getCreatedAt)), Optional::get)
+                ));
+
+
+        var agreementsById = getAllContractAgreements().stream().collect(toMap(
+                ContractAgreement::getId, Function.identity()
+        ));
+
+        var assetsById = getAllAssets().stream()
+                .collect(toMap(Asset::getId, Function.identity()));
+
+        var transferStream = getAllTransferProcesses().stream();
```

Comment:
don't assign streams to variables, if it was

```java
var transferProcesses = getAllTransferProcesses();
return tranfserProcesses.stream().map(process -> {
    // ...
});
```

it would be more reasonable